scapy/utils.py

```diff
@@ -1365,33 +1365,78 @@ def __init__(self, filename, fdesc, magic):
             "tsresol": 1000000
         }
         self.blocktypes = {
-            1: self.read_block_idb,
-            2: self.read_block_pkt,
-            3: self.read_block_spb,
-            6: self.read_block_epb,
+            1: self._read_block_idb,
+            2: self._read_block_pkt,
+            3: self._read_block_spb,
+            6: self._read_block_epb,
         }
+        self.endian = "!"  # Will be overwritten by first SHB
+
         if magic != b"\x0a\x0d\x0d\x0a":  # PcapNg:
             raise Scapy_Exception(
                 "Not a pcapng capture file (bad magic: %r)" % magic
             )
-        # see https://github.com/pcapng/pcapng
-        blocklen_, magic = self.f.read(4), self.f.read(4)  # noqa: F841
-        if magic == b"\x1a\x2b\x3c\x4d":
+
+        try:
+            self._read_block_shb()
+        except EOFError:
+            raise Scapy_Exception(
+                "The first SHB of the pcapng file is malformed !"
+            )
+
+    def _read_block(self, size=MTU):
+        # type: (int) -> Optional[Tuple[bytes, RawPcapNgReader.PacketMetadata]]  # noqa: E501
+        try:
+            blocktype = struct.unpack(self.endian + "I", self.f.read(4))[0]
+        except struct.error:
+            raise EOFError
+        if blocktype == 0x0A0D0D0A:
+            # This function updates the endian based on the block content.
+            self._read_block_shb()
+            return None
+        blocklen = struct.unpack(self.endian + "I", self.f.read(4))[0]
+        if blocklen < 12:
+            warning("Invalid block length !")
+            raise EOFError
+        block = self.f.read(blocklen - 12)
+        self._read_block_tail(blocklen)
+        return self.blocktypes.get(
+            blocktype,
+            lambda block, size: None
+        )(block, size)
+
+    def _read_block_tail(self, blocklen):
```

Comment:
IMHO, we must check all calls to read to ensure that they do not fail.

Reply:
I don't see how those calls can fail. In my experience, either they fail on first read (this is handled on init), either they return an empty string which eventually triggeres a `struct.error`

Reply:
I think that each of these functions shall do its best to check that they won't crash. AFAIR, the bugs found via oss-fuzz were related to missing bound checks.

Reply:
Regarding https://github.com/secdev/scapy/pull/3047
Do you think it's better to return an empty pcap than to raise an error ?

Reply:
IMHO, an empty packelist is better than an exception as long as we also display a message.

Reply:
I'd like to disagree. `Scapy_Exception`s are only triggered when the whole pcap file is invalid (and no packet can be read. e.g. wrong header length..), while `EOFError`s are triggered when a single packet/packet block is malformed.
`EOFError`s are caught in `read_all` or `sniff`, which will return whatever has been dissected so far, while `Scapy_Exceptions` are not caught anywhere and returned to the user.
Considering many users disable logs, I think there's a level of criticalness in raising an error that we should keep.

Reply:
I am fine with the current behavior.